tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala

```diff
@@ -0,0 +1,121 @@
+
+  private def handleError(publicAssetCreationId: PublicAssetCreationId, error: Throwable): SMono[PublicAssetCreationResult] =
+    error match {
+      case e: PublicAssetException => SMono.just(PublicAssetCreationFailure(publicAssetCreationId, e))
+      case e => throw e
+    }
+
+  private def createPublicAssets(mailboxSession: MailboxSession,
+                                 creationRequest: PublicAssetCreationRequest,
+                                 accountId: AccountId): SMono[PublicAsset] =
+    generatePublicAsset(creationRequest, mailboxSession, accountId)
+      .flatMap(publicAsset => SMono(publicAssetRepository.create(mailboxSession.getUser, publicAsset)))
+
+  private def parseCreate(jsObject: JsObject): Either[PublicAssetCreationParseException, PublicAssetCreationRequest] =
+    PublicAssetCreationRequest.validateProperties(jsObject)
+      .flatMap(validJsObject => PublicAssetSerializer.deserializePublicAssetCreationRequest(validJsObject) match {
+        case JsSuccess(creationRequest, _) => Right(creationRequest)
+        case JsError(errors) => Left(PublicAssetCreationParseException(standardError(errors)))
+      })
+
+  private def generatePublicAsset(creationRequest: PublicAssetCreationRequest,
+                                  session: MailboxSession,
+                                  accountId: AccountId): SMono[PublicAsset] =
+    for {
+      parsedIdentityIds <- parseIdentityIds(creationRequest)
+      uploadId = publicAssetSetService.uploadIdFromBlobId(creationRequest.blobId).fold(e => throw e, identity)
+      identityIds <- publicAssetSetService.checkIdentityIdsExist(parsedIdentityIds, session)
+      uploadMetadata <- getUploadMetadata(session, uploadId)
+      publicAssetId: PublicAssetId = PublicAssetIdFactory.generate()
+      publicURI: PublicURI = publicAssetSetService.generatePublicURI(publicAssetId, accountId)
+    } yield {
+      PublicAsset(
+        id = publicAssetId,
+        publicURI = publicURI,
+        size = uploadMetadata.size,
+        contentType = ImageContentType.from(uploadMetadata.contentType),
+        identityIds = identityIds,
+        blobId = publicAssetSetService.blobIdFromUploadId(uploadId).get)
+    }
+
+  private def getUploadMetadata(session: MailboxSession, uploadId: UploadId): SMono[UploadMetaData] =
+    SFlux(uploadRepository.listUploads(session.getUser)) // it should be better if uploadRepository support method to get upload by uploadId
+      .filter(_.uploadId.equals(uploadId))
+      .next()
+      .switchIfEmpty(SMono.error(PublicAssetBlobIdNotFoundException(publicAssetSetService.blobIdFromUploadId(uploadId).get.value.value)))
+
+  private def parseIdentityIds(creationRequest: PublicAssetCreationRequest): SMono[Seq[IdentityId]] =
+    creationRequest.identityIds match {
+      case None => SMono.just(Seq.empty)
+      case Some(value) if value.ids.isEmpty => SMono.just(Seq.empty)
+      case Some(value) =>
+        SFlux.fromIterable(value.ids)
```

Comment:
As far as I understood nothing is asynchronous as it is only validation. Then please favor `Either` to represent for potential failures.